### PR TITLE
Adding Olympus benchmark

### PR DESCRIPTION
**Pull request by ncarraz**
## Types of changes

- [x] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] The documentation is up-to-date with the changes I made.
- [x] I have read the [**CONTRIBUTING**](https://facebookresearch.github.io/nevergrad/contributing.html) document and completed the CLA (see [**CLA**](https://facebookresearch.github.io/nevergrad/contributing.html#contributor-license-agreement-cla)).
- [ ] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by ncarraz** [comment]
Benchmark on the traditional surfaces only. I did not include Branin because there was an internal Olympus error when trying to instatiate it. No noise yet but to be included later.

File: nevergrad/functions/olympus/core.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import numpy as np
+from nevergrad.parametrization import parameter as p
+from ..base import ExperimentFunction
+import olympus
+
+
+class Olympus(ExperimentFunction):
+    def __init__(self, kind : str, dimension: int = 10) -> None:
+        self.kind = kind
+        self.param_dim = dimension
```
Comment (teytaud) [inline nevergrad/functions/olympus/core.py:18]: This is correct but does not inform the optimization algorithm that the objective function is noisy, I would do something like

parametrization = p.Array(shape=(dimension,))
parametrization.function.deterministic = False

(@jrapin , big boss of parametrization stuff, might know more than i do :-) )

File: nevergrad/functions/olympus/core.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        super().__init__(self._simulate_traditional_surface, p.Array(shape=(dimension,)))
+    
+    def _simulate_traditional_surface(self, x: np.ndarray) -> float:
+        assert self.kind in ["AckleyPath", "Dejong", "HyperEllipsoid",
+                        "Levy", "Michalewicz", "Rastrigin", "Rosenbrock", 
+                        "Schwefel", "StyblinskiTang", "Zakharov"]
+
```
Comment (teytaud) [inline nevergrad/functions/olympus/core.py:25]: Sounds correct to me, but for making assessment faster we can average noise by specifying that for a posteriori evaluation we might average 42 runs as follows (this does not impact the run, just the evaluation of the final solution at the end...):

def evaluation_function(self, *recommendations) -> float:
        """Averages multiple evaluations if necessary."""
        x = recommendations[0].value
        losses = [surface.run(x)[0][0] for _ in range(42)]
        return sum(losses) / len(losses)

**Comment by teytaud** [comment]
> Benchmark on the traditional surfaces only. I did not include Branin because there was an internal Olympus error when trying to instatiate it. No noise yet but to be included later.

ah sorry for having commented without seeing that you had already mentioned that noise would come up later :-)

File: nevergrad/functions/olympus/core.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/functions/olympus/core.py:25]: Even better (if possible) would be to have "def evaluation_function" return the value without noise (while still using the noisy one in super.__init__(....., ...) ).

**Comment by ncarraz** [comment]
test_olympus in test_core.py does not test everything yet. I managed to use the surface without the noise in evaluation_function but you should check.

**Comment by ncarraz** [comment]
I think it passed the tests but I run out of credits in circkeci

File: nevergrad/benchmark/experiments.py [hunk 3]
```diff
@@ -1233,6 +1234,39 @@ def neuro_control_problem(seed: tp.Optional[int] = None) -> tp.Iterator[Experime
...
+        "HyperEllipsoid",
+        "Levy",
+        "Michalewicz",
+        "Rastrigin",
+        "Rosenbrock",
+        "Schwefel",
+        "StyblinskiTang",
+        "Zakharov",
+    ]:
+        for k in range(2, 6):
```
Comment (teytaud) [inline nevergrad/benchmark/experiments.py:1253]: Maybe 5 is enough because dim = 1 000 000 in a noisy context will be a lot.

Comment (ncarraz) [reply nevergrad/benchmark/experiments.py:1253]: okay I will fix that. what about the noise scale ?

File: nevergrad/functions/olympussurfaces/core.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            "Schwefel": surfaces.Schwefel,
+            "StyblinskiTang": surfaces.StyblinskiTang,
+            "Zakharov": surfaces.Zakharov,
+            "DiscreteAckley": surfaces.DiscreteAckley,
+            "DiscreteDoubleWell": surfaces.DiscreteDoubleWell,
+            "DiscreteMichalewicz": surfaces.DiscreteMichalewicz,
+            "LinearFunnel": surfaces.LinearFunnel,
+            "NarrowFunnel": surfaces.NarrowFunnel,
+            "GaussianMixture": surfaces.GaussianMixture,
+        }
```
Comment (jrapin) [inline nevergrad/functions/olympussurfaces/core.py:70]: Please move this to a global variable, then you can create a function `get_kinds()` which returns `list(TRADITIONAL_SURFACES)`
This will remove a lot of duplicated items, and therefore will be simpler to maintain

File: nevergrad/functions/olympussurfaces/core.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+            surface = traditional_surfaces[self.kind](param_dim=self.param_dim, noise=noise)
+        else:
+            surface = traditional_surfaces[self.kind](param_dim=self.param_dim)
+        return surface.run(x)[0][0]
+
+    def evaluation_function(self, *recommendations) -> float:
+        """Averages multiple evaluations if necessary."""
+        x = recommendations[0].value
+        losses = [self.surface_without_noise(x) for _ in range(42)]
+        return sum(losses) / len(losses)
```
Comment (jrapin) [inline nevergrad/functions/olympussurfaces/core.py:82]: would that work to use the non-noisy case instead?
```python
traditional_surfaces[self.kind](param_dim=self.param_dim)
return surface.run(x)[0][0]
```

Comment (ncarraz) [reply nevergrad/functions/olympussurfaces/core.py:82]: Using the non-noisy surface in the evaluation function was proposed by Olivier.

Comment (teytaud) [reply nevergrad/functions/olympussurfaces/core.py:70]: Yes! But now we can remove the loop.

Comment (ncarraz) [reply nevergrad/functions/olympussurfaces/core.py:70]: ok got it

File: nevergrad/functions/olympussurfaces/core.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+
+import nevergrad as ng
+
+try:
+    from olympus import surfaces  # type: ignore
+    from olympus import noises  # type: ignore
+    from olympus import Emulator  # type: ignore
+    from olympus.datasets import Dataset  # type: ignore
+except ImportError as e:
+    raise ng.errors.UnsupportedExperiment("Please install olympus for Olympus experiments") from e
```
Comment (jrapin) [inline nevergrad/functions/olympussurfaces/core.py:21]: This needs to be done inline in the function, otherwise it will break all other experiments.

Also, instead of all the type ignore, add `olympus` to mypy.ini:
https://github.com/facebookresearch/nevergrad/blob/603221a535dfaebfd7d76b56732cbd9b716292eb/mypy.ini#L3

File: nevergrad/functions/olympussurfaces/core.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        "Schwefel": surfaces.Schwefel,
+        "StyblinskiTang": surfaces.StyblinskiTang,
+        "Zakharov": surfaces.Zakharov,
+        "DiscreteAckley": surfaces.DiscreteAckley,
+        "DiscreteDoubleWell": surfaces.DiscreteDoubleWell,
+        "DiscreteMichalewicz": surfaces.DiscreteMichalewicz,
+        "LinearFunnel": surfaces.LinearFunnel,
+        "NarrowFunnel": surfaces.NarrowFunnel,
+        "GaussianMixture": surfaces.GaussianMixture,
+    }
```
Comment (jrapin) [inline nevergrad/functions/olympussurfaces/core.py:44]: can this be created with just: `{name: getattr(surfaces, name) for name in dir(surfaces)}`?
There may be a few additional gotchas (getting only if it's the correct class, skipping some of the names which are not supported) but this way the code will be much more maintainable.

Or otherwise just create `names = ["Michalewicz", ...]` if you prefer to make sure we do not use other surfaces, then, the corresponding object is just `getattr(surfaces, name)`, so we can remove a lot of duplications

File: nevergrad/functions/olympussurfaces/core.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        "alkox",
+        "benzylation",
+        "photo_pce10",
+        "hplc",
+        "colors_n9",
+    ]
+
+    @classmethod
+    def get_datasets(cls):
+        return cls.datasets
```
Comment (jrapin) [inline nevergrad/functions/olympussurfaces/core.py:97]: Simpler to make it explicitely an unmutable (= a tuple instead of a list) class attribute, and use it directly afterwards

```suggestion
    DATASETS = (
        "suzuki",
        "fullerenes",
        "colors_bob",
        "photo_wf3",
        "snar",
        "alkox",
        "benzylation",
        "photo_pce10",
        "hplc",
        "colors_n9",
    )
```

**Review by jrapin — COMMENTED** [review]
Hi,
See a few comments here
Please merge the latest master within your branch, this should simplify your PR a bit, and solve the current pytest issues which are unrelated to your PR

File: nevergrad/benchmark/test_experiments.py [hunk 9]
```diff
@@ -36,8 +36,8 @@ def test_experiments_registry(name: str, maker: tp.Callable[[], tp.Iterator[expe
     if "mixsimulator" in name and os.environ.get("CIRCLECI", False):
         raise SkipTest("Sigkill in CircleCI")
 
-    # Our IQAs and our ScikitLearn are not well guaranteed on Windows.
-    if all(x in name for x in ["image", "quality"]) and platform.system() == "Windows":
+    # Our IQAs and our ScikitLearn and Olympus are not well guaranteed on Windows.
+    if all(x in name for x in ["image", "quality", "olympus"]) and platform.system() == "Windows":
```
Comment (jrapin) [inline nevergrad/benchmark/test_experiments.py:40]: this should not be necessary anymore now that it raises UnsupportedExperiment, so please try and revert the change

File: nevergrad/functions/olympussurfaces/core.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+        return p.Array(shape=(dimension,), lower=bounds[0], upper=bounds[1])
+
+    def _simulate_emulator(self, x: np.ndarray) -> float:
+        try:
+            from olympus import Emulator  # pylint: disable=import-outside-toplevel
+        except ImportError as e:
+            raise ng.errors.UnsupportedExperiment("Please install olympus for Olympus experiments") from e
+
+        assert self.dataset_kind in OlympusEmulator.DATASETS
+        assert self.model_kind in ["BayesNeuralNet", "NeuralNet"]
```
Comment (jrapin) [inline nevergrad/functions/olympussurfaces/core.py:116]: test this in the `__init__` instead, no need to have it tested at each call of the function

File: nevergrad/functions/olympussurfaces/core.py [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+
+        emulator = Emulator(dataset=self.dataset_kind, model=self.model_kind)
+        if emulator.get_goal() == "maximize":
+            return -emulator.run(x)[0][0]
+        else:
+            return emulator.run(x)[0][0]
```
Comment (jrapin) [inline nevergrad/functions/olympussurfaces/core.py:122]: I tend to factorize all repeated code asap, since fewer repetition usually means less maintenance :)

```suggestion
        return emulator.run(x)[0][0] * (-1 if emulator.get_goal() == "maximize" else 1)
```

File: requirements/bench.txt [hunk 12]
```diff
@@ -26,3 +26,5 @@ tensorflow>=2.5.0
 image-quality>=1.2.7
 keras>=2.4.3
 compiler_gym>=0.1.8 ; sys_platform == "linux"
+olymp==0.0.1b0 ; sys_platform == "linux"
+silence_tensorflow
```
Comment (jrapin) [inline requirements/bench.txt:30]: there seems to be a few more missing imports (cf CI)

**Review by jrapin — COMMENTED** [review]
looks like we're nearly there ;)

File: requirements/bench.txt [hunk 13]
```diff
@@ -29,3 +29,8 @@ keras>=2.4.3
 compiler_gym>=0.1.8 ; sys_platform == "linux"
 pymoo>=0.4.2.2
 bayes-optim>=0.2.5.5
+olymp==0.0.1b0 ; sys_platform == "linux"
+silence_tensorflow
+tensorflow_probability
+sqlalchemy
+seaborn
```
Comment (jrapin) [inline requirements/bench.txt:36]: are those really needed? looks like database and plotting tools so that feels unrelated

**Review by jrapin — COMMENTED** [review]
I have no idea what is the problem with the test failing in the CI, I have tried updating all the dependency to check if it was due to a dependency update but all worked (#1220), so it must be due to one of the additional imports, no idea which though :s and as long as the CI does not pass we won't be able to merge :s

File: requirements/bench.txt (hunk 13 above)
Comment (ncarraz) [reply requirements/bench.txt:36]: Those are used internaly by Olympus. I could try removing them and see if there are no issues.

Comment (ncarraz) [reply requirements/bench.txt:36]: It was Koncept512 that failed in pytests. Is it possible that it is related ?

Comment (ncarraz) [reply requirements/bench.txt]: It works locally though.

Comment (jrapin) [reply requirements/bench.txt]: > It was Koncept512 that failed in pytests. Is it possible that it is related ?

Initially I would have expected it to be unrelated, but after a clean install of master on the CI, this issue does not appear. So what triggers the problem is necessaritly somewhere here, I would expect one of the packages, and because koncept512 is in tensorflow, probably silence_tensorflow or tensorflow_probability. tensorflow version and keras version are the same in master though so it's not "simply" a problem version :s
Is silence_tensorflow needed? Might be good to try without it I guess :s

File: nevergrad/benchmark/test_experiments.py [hunk 14]
```diff
@@ -55,7 +55,7 @@ def test_experiments_registry(name: str, maker: tp.Callable[[], tp.Iterator[expe
         maker,
         ("mltuning" in name or "anm" in name),
         skip_seed=(name in ["rocket", "images_using_gan"])
-        or any(x in name for x in ["tuning", "image_", "compiler", "anm"]),
+        or any(x in name for x in ["tuning", "image_", "compiler", "anm", "olympus"]),
```
Comment (jrapin) [inline nevergrad/benchmark/test_experiments.py:58]: is olympus not seedable in some way? If you had to add this here it means the test cases would not be reproducible :(
I see the shift is using the parametrization random state which is seeded so should be good, so that must come from olympus

Comment (ncarraz) [reply nevergrad/benchmark/test_experiments.py:58]: Unfortunately, I have not found any ways to seed the emulators but the surfaces are seedable.

File: requirements/bench.txt (hunk 13 above)
Comment (ncarraz) [reply requirements/bench.txt]: I will have to check all that.

File: nevergrad/benchmark/test_experiments.py (hunk 14 above)
Comment (jrapin) [reply nevergrad/benchmark/test_experiments.py:58]: too bad :(
